### ISSUE TITLE
consul rpc error lead fd leaks

### ISSUE THREAD

**Issue by hehailong5** [hashicorp/consul#8312]
consul version: v1.6.6

my app uses consul apis to talk with the consul server, when the below errors appear in the consul log,

2020/07/15 13:54:49 [ERR] http: Request GET /v1/kv/xxxx/xxxxx/xxxx/xxxx?index=8323460&wait=600000ms, error: rpc error making call: EOF from=172.20.0.4:41790
2020/07/15 13:54:49 [ERR] http: Request GET /v1/kv/xxxx/xxxx?index=7391853&recurse=&wait=600000ms, error: rpc error making call: EOF from=172.20.0.4:51744
2020/07/15 13:54:49 [ERR] http: Request GET /v1/health/service/xxxxx?index=8247150&wait=600000ms, error: rpc error making call: EOF from=172.20.0.4:52366

there are lots "sock" entries in the resulf of "lsof -p" and never go away,

xxxxx 43629  xx 1418u     sock        0,7      0t0 1266922513 protocol: TCP
xxxxx 43629  xx 1419u     sock        0,7      0t0 1239538922 protocol: TCP
xxxxx 43629  xx 1420u     sock        0,7      0t0 1268874931 protocol: TCP
xxxxx 43629  xx 1421u     sock        0,7      0t0 1446939411 protocol: TCP
xxxxx 43629  xx 1422u     sock        0,7      0t0 1472627582 protocol: TCP

**Comment by mkeeler** [comment]
How many are "lots". Every server will keep 1 persistent connection to all other servers and each client will keep 1 persistent connection open one server (this gets changed roughly every 5 minutes where a client "rotates" which server its using to send RPCs). Beyond just those there are other sockets open for HTTP requests and if you are using the consul api client these will be pooled with keep-alives to ensure the connection remains open.

Thats not to say that a socket leak isn't impossible but rather it would be good to get an idea of just how many sockets are hanging around.

Additionally, providing some more information about your cluster (how many servers, how many clients, how many datacenters etc.) could help in determining whats going on.

**Comment by hehailong5** [comment]
Hi,
Our app connects to one consul server of a three nodes cluster via its http port.
There are about 30 different queries (some of them are blocking queries) repeated 8 thousand times with this rpc error, and leave around 1500 "sock" entries in the lsof result.
Please let me know if more info is needed.
Thank you!

**Comment by bretthoerner** [comment]
I'm not the OP, but it seems like we have a related issue.

* Consul 1.6.3
* 5 Consul servers
* One datacenter has 260 servers (each running a consul-agent) and 1 of those has leaked and maxed out on fds, leaving the agent in a broken state for new clients
* Another datacenter has 340 servers and approximately 5 of those leaked enough fds to get into the bad state, we restarted those last week when we noticed and they have been fine since
* The only configuration that seems pertinent is `http_max_conns_per_client = 1000`
* These broken nodes hit ~1014 fds and are then effectively busted for any new clients
* Healthy consul-agents in our datacenters have about 20-25 fds open
* We only connect to the consul-agent on localhost and all of our servers run the ~same software
* The vast majority of the fds in `lsof` list `protocol: TCP` just like in OPs report

`lsof` output: https://gist.githubusercontent.com/bretthoerner/a72dc5faee0a712147ce32a85a22d49c/raw/68a6504ae704d1a661d6754ddce364adedfce143/consul-lsof.txt

I didn't see anything in the release notes that looked like this had been known/fixed, or we'd prioritize an upgrade.

My coworker pointed out the `protocol: TCP` with no source/destination smell like a classic Go leak: https://github.com/golang/go/issues/20413

What other information can I collect for y'all?

**Comment by a-robinson** [comment]
A similar symptom of a consul agent getting stuck was seen in https://github.com/hashicorp/consul/issues/9462, although that case is slightly different because the agent itself didn't leak FDs, it was the system that was out of available FDs temporarily. Even once FDs were available again the consul agent refused to accept new connections.

**Comment by markblackman** [comment]
https://github.com/hashicorp/consul/blob/cdcfd0e4f95afd9f6d868b0e2fbed3044b41975e/agent/checks/check.go#L584 looks like it might be leaking FDs on failed TCP healthchecks. I will test this hypothesis this week, but thought I would leave a note for someone who knows if there's a mitigating code path elsewhere. I would guess this should be a `defer conn.Close()` right after the socket.

**Comment by markblackman** [comment]
I cannot reproduce the issue simply by specifying a nonexistent endpoint for a TCP healthcheck, so the issue is more subtle than something as simple as that, which I expected it would be. It's trivial to reproduce the 'protocol: TCP' state in a bit of C code but I cannot yet come up with any Go version that will result in this case, using net.Dialer, although I would imagine calling syscalls directly should reproduce it in Go.

**Comment by markblackman** [comment]
I have reproduced the case where a net.Dialer() code will result in sockets without connections. As expected, this is as simple as starting a client connection to a server, then killing the server side, i.e. the other end of the connection. The socket spends around 60 seconds in CLOSE_WAIT, then progresses to an unconnected socket. I assume the reverse is also true, a client disconnecting unexpectedly from a server that doesn't scrupulously handle all connection closure cases, will also leave unconnected sockets unclosed. I am speculating we have clients that are disconnecting from the consul agent HTTP interface unexpectedly, but that the consul agent HTTP server isn't closing the socket as expected. I have skimmed the code for obvious problems, but not spotted any.

**Comment by mkeeler** [comment]
@markblackman I think you are correct in that we should defer the `conn.Close()` so that it is performed even in the error case.

**Comment by mkeeler** [comment]
Took a deeper look at the `net.Dialers` `Dial` method and now I don't think deferring `conn.Close` would help. All paths in the go source either return a nil Conn or a nil error. There is no path that I have seen that will return a valid conn with an error.

**Comment by markblackman** [comment]
@mkeeler Agreed, and thinking about it a bit longer, I concluded these were actually sockets that had been fully disconnected (exited CLOSE_WAIT) but not yet actually closed which is obviously a more likely case.  I did some digging in our case, and it looks even odder. We believe the trigger are big processes getting OOM-killed on one of the consul peer servers and the subsequent  OOM-processing takes ~50 seconds and during this time, the kernel is pretty unresponsive and this triggers a a lot of instability in the raft peer servers and somehow this results in a small number of consul client agents with these disconnected-but-open sockets. Any theories?
